beginend.el

```diff
@@ -61,107 +62,165 @@
+       (defun ,beginfunc-name ()
+         ,(format "Go to beginning of buffer in `%s'." mode-name)
+         (interactive)
+         (beginend--double-tap-begin
+          ,@(cdr begin-body)))
+       (defun ,endfunc-name ()
+         ,(format "Go to beginning of buffer in `%s'." mode-name)
+         (interactive)
+         (beginend--double-tap-end
+          ,@(cdr end-body)))
+       (defvar ,map-name
+         (let ((map (make-sparse-keymap)))
+           (beginend--defkey map #',beginfunc-name #',endfunc-name)
+           map)
+         ,(format "Keymap for beginend in `%s'." mode-name))
+       (define-minor-mode ,beginend-mode-name
+         ,(format "Mode for beginend in `%s'.\n\\{%s}" mode-name map-name)
+         :lighter " be"
+         :keymap ,map-name)
+       (add-to-list 'beginend--modes
+                    (cons ',hook
+                          #',beginend-mode-name)))))
+
+;;; Modes
+
+(beginend-define-mode message-mode
+  (progn
+    (message-goto-body))
+  (progn
+    (when (re-search-backward "^-- $" nil t)
+      (beginend--goto-nonwhitespace))))
+
+(beginend-define-mode dired-mode
+  (progn
+    (let ((move 4))
+      (when (and (boundp 'dired-omit-mode) dired-omit-mode)
+        ;; dired-omit-mode hides `.' and `..'.
+        (setf move (- move 2)))
+      (when (and (boundp 'dired-hide-details-hide-information-lines)
+                 dired-hide-details-hide-information-lines
+                 (boundp 'dired-hide-details-mode)
+                 dired-hide-details-mode)
+        ;; 1 line containing directory size
+        (setf move (- move 1)))
+      (dired-next-line move)))
+  (progn
+    (beginend--goto-nonwhitespace)))
+
+(beginend-define-mode occur-mode
+  (progn
+    (occur-next 1))
+  (progn
+    (occur-prev 1)))
+
+(beginend-define-mode ibuffer-mode
+  (progn
+    (ibuffer-forward-line 1))
+  (progn
+    (ibuffer-backward-line 1)))
+
+(beginend-define-mode vc-dir-mode
+  (progn
+    (vc-dir-next-line 1))
+  (progn
+    (vc-dir-previous-line 1)))
+
+(beginend-define-mode bs-mode
+  (progn
+    (bs-down 2))
+  (progn
+    (bs-up 1)
+    (bs-down 1)))
+
+(beginend-define-mode recentf-dialog-mode
+  (progn
+    (when (re-search-forward "^  \\[" nil t)
+      (goto-char (match-beginning 0))))
+  (progn
+    (re-search-backward "^  \\[" nil t)))
+
+(beginend-define-mode org-agenda-mode
+  (progn
+    (org-agenda-next-item 1))
+  (progn
+    (org-agenda-previous-item 1)))
+
+(beginend-define-mode compilation-mode
+  (progn
+    (compilation-next-error 1))
+  (progn
+    (compilation-previous-error 1)))
+
+(beginend-define-mode notmuch-search-mode
+  (progn
+    (notmuch-search-first-thread)
+    (beginning-of-line))
+  (progn
+    (notmuch-search-last-thread)
+    (end-of-line)))
+
+(beginend-define-mode elfeed-search-mode
+  (progn)
+  (progn
+    (forward-line -2)))
+
+(beginend-define-mode prodigy-mode
+  (progn
+    (prodigy-first))
+  (progn
+    (prodigy-last)))
 
 ;;;###autoload
 (defun beginend-setup-all ()
```

Comment:
Maybe this could be turned into a minor mode as well (a global one?) that would just update/remove the hooks when activated/deactivated.  That way I can toggle it on in customize and never care to call some code in my config, it would happen automatically during init.

In addition to that I also have the fine control of every separate minor mode for each buffer, which is nice.